lib/puppet/forge/repository.rb

```diff
@@ -15,38 +15,52 @@ class Repository
 
     attr_reader :uri, :cache
 
-    # List of Net::HTTP exceptions to catch
-    NET_HTTP_EXCEPTIONS = [
-      EOFError,
-      Errno::ECONNABORTED,
-      Errno::ECONNREFUSED,
-      Errno::ECONNRESET,
-      Errno::EINVAL,
-      Errno::ETIMEDOUT,
-      Net::HTTPBadResponse,
-      Net::HTTPHeaderSyntaxError,
-      Net::ProtocolError,
-      SocketError,
-    ]
-
-    if Puppet.features.zlib?
-      NET_HTTP_EXCEPTIONS << Zlib::GzipFile::Error
-    end
-
     # Instantiate a new repository instance rooted at the +url+.
     # The library will report +for_agent+ in the User-Agent to the repository.
     def initialize(host, for_agent)
       @host  = host
       @agent = for_agent
       @cache = Cache.new(self)
       @uri   = URI.parse(host)
+
+      ssl_provider = Puppet::SSL::SSLProvider.new
+      @ssl_context = ssl_provider.create_system_context(cacerts: [])
     end
 
     # Return a Net::HTTPResponse read for this +path+.
     def make_http_request(path, io = nil)
-      request = get_request_object(@uri.path.chomp('/')+path)
-      Puppet.debug "HTTP GET #{@host}#{request.path}"
-      return read_response(request, io)
+      raise ArgumentError, "Path must start with forward slash" unless path.start_with?('/')
+      begin
+        str = @uri.to_s
+        str.chomp!('/')
+        str += path
+        uri = URI(Puppet::Util.uri_encode(str))
+
+        headers = { "User-Agent" => user_agent }
+        user = nil
+        password = nil
+
+        if forge_authorization
+          headers["Authorization"] = forge_authorization
+        elsif @uri.user && @uri.password
+          user = @uri.user
+          password = @uri.password
+        end
+
+        http = Puppet.runtime['http']
+        response = http.get(uri, headers: headers, user: user, password: password, ssl_context: @ssl_context)
+        io.write(response.body) if io.respond_to?(:write)
+        response
+      rescue => e
+        case e.cause
+        when Puppet::SSL::CertVerifyError
+          raise SSLVerifyError.new(:uri => @uri.to_s, :original => e.cause)
+        when Puppet::SSL::SSLError
+          raise e.cause
```

Comment:
This seems a little suspect. The http client API only raises `Puppet::HTTP::HTTPErrors` and derived exceptions. But here we need to differentiate between 1) cert verification error, 2) general ssl error, 3) non-ssl error. Peeking at the cause is cheating, as it's not part of the API. We could:

* Have the client raise `Puppet::SSL::SSLErrors` in addition to `Puppet::HTTP::HTTPError` and subclasses
* Create `Puppet::HTTP::SSLError` to mirror their `Puppet:SSL` counterparts

I'm thinking the former is better, but tests need to be added.

Reply:
I implemented the former in https://github.com/puppetlabs/puppet/pull/7864/commits/aecceef1eef9afc9030585870560902d3776acff, but would be good to discuss.